src/v/cloud_storage/manifest.cc

```diff
@@ -33,10 +34,208 @@
 
 #include <algorithm>
 #include <array>
+#include <charconv>
 #include <chrono>
+#include <variant>
 
 namespace cloud_storage {
 
+std::ostream& operator<<(std::ostream& s, const manifest_path_components& c) {
+    fmt::print(
+      s, "{{{}: {}-{}-{}-{}}}", c._origin, c._ns, c._topic, c._part, c._rev);
+    return s;
+}
+
+std::ostream& operator<<(std::ostream& s, const segment_path_components& c) {
+    fmt::print(
+      s,
+      "{{{}: {}-{}-{}-{}-{}}}",
+      c._origin,
+      c._ns,
+      c._topic,
+      c._part,
+      c._rev,
+      c._name);
+    return s;
+}
+
+static bool parse_partition_and_revision(
+  std::string_view s, manifest_path_components& comp) {
+    auto pos = s.find('_');
+    if (pos == std::string_view::npos) {
+        // Invalid segment file name
+        return false;
+    }
+    uint64_t res = 0;
+    // parse first component
+    auto sv = s.substr(0, pos);
+    auto e = std::from_chars(sv.data(), sv.data() + sv.size(), res);
+    if (e.ec != std::errc()) {
+        return false;
+    }
+    comp._part = model::partition_id(res);
+    // parse second component
+    sv = s.substr(pos + 1);
+    e = std::from_chars(sv.data(), sv.data() + sv.size(), res);
+    if (e.ec != std::errc()) {
+        return false;
+    }
+    comp._rev = model::revision_id(res);
+    return true;
+}
+
+std::optional<manifest_path_components>
+get_manifest_path_components(const std::filesystem::path& path) {
+    // example: b0000000/meta/kafka/redpanda-test/4_2/manifest.json
+    enum {
+        ix_prefix,
+        ix_meta,
+        ix_namespace,
+        ix_topic,
+        ix_part_rev,
+        ix_file_name,
+        total_components
+    };
+    manifest_path_components res;
+    res._origin = path;
+    int ix = 0;
+    for (const auto& c : path) {
+        ss::sstring p = c.string();
+        switch (ix++) {
+        case ix_prefix:
+            break;
+        case ix_namespace:
+            res._ns = model::ns(std::move(p));
+            break;
+        case ix_topic:
+            res._topic = model::topic(std::move(p));
+            break;
+        case ix_part_rev:
+            if (!parse_partition_and_revision(p, res)) {
+                return std::nullopt;
+            }
+            break;
+        case ix_file_name:
+            if (p != "manifest.json") {
+                return std::nullopt;
+            }
+            break;
+        }
+    }
+    if (ix == total_components) {
+        return res;
+    }
+    return std::nullopt;
+}
+
+/// Parse segment file name
+/// \return offset and success flag
+std::tuple<model::offset, bool>
+get_base_offset(const std::filesystem::path& path) {
+    static constexpr auto bad_result = std::make_tuple(model::offset(), false);
```

Comment:
Return value could be an optional rather a special error value.

However I was looking for where this is used and it seems to not be?